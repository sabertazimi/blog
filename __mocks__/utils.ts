let navigatorClipboardMocked = false
const writeTextMock = jest.fn()

const mockNavigatorClipboard = () => {
  if (!navigatorClipboardMocked) {
    Object.defineProperty(global.navigator, 'clipboard', {
      value: {
        writeText: writeTextMock,
      },
    })

    navigatorClipboardMocked = true
  }

  return { writeTextMock }
}

export { mockNavigatorClipboard }
