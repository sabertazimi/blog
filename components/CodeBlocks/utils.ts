import parseNumericRange from 'parse-numeric-range'
import type { Language } from 'prism-react-renderer'

const normalizeCode = (code: string = '') => code.replace(/\n+$/, '')

function isLanguage(language: string): language is Language {
  return language.length > 0
}

function normalizeLanguage(languageClass?: string): Language {
  if (languageClass != null) {
    const language = languageClass.replace('language-', '')

    if (isLanguage(language))
      return language
  }

  return 'typescript'
}

function normalizeLanguageName(language: string): string {
  switch (language) {
    case 'html':
      return 'HTML'
    case 'xml':
      return 'XML'
    case 'yml':
    case 'yaml':
      return 'YAML'
    case 'css':
      return 'CSS'
    case 'json':
      return 'JSON'
    case 'md':
    case 'markdown':
      return 'Markdown'
    case 'js':
    case 'javascript':
      return 'JavaScript'
    case 'ts':
    case 'typescript':
      return 'TypeScript'
    case 'coffee':
    case 'coffeescript':
      return 'CoffeeScript'
    case 'jsx':
    case 'tsx':
      return 'React'
    case 'objc':
    case 'objectivec':
      return 'Objective-C'
    default:
      return language.charAt(0).toUpperCase() + language.slice(1)
  }
}

function normalizeLines(expression: string): Set<number> {
  return new Set(parseNumericRange(expression))
}

export {
  normalizeCode,
  normalizeLanguage,
  normalizeLanguageName,
  normalizeLines,
}
