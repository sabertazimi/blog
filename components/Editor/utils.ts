function languageToFilepath(language?: string): string {
  switch (language) {
    case 'css':
      return '/styles.css'
    case 'js':
    case 'javascript':
      return '/App.js'
    case 'ts':
    case 'typescript':
      return '/App.ts'
    case 'jsx':
      return '/App.jsx'
    case 'tsx':
      return '/App.tsx'
    case 'vue':
      return '/src/App.vue'
    case undefined:
    default:
      return '/App.tsx'
  }
}

function normalizeFilepath(filename?: string, language?: string): string {
  if (filename == null || filename === '')
    return languageToFilepath(language)

  if (!filename.startsWith('/'))
    return `/${filename}`

  return filename
}

export { normalizeFilepath }
