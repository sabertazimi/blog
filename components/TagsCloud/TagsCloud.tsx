import { LinkTag } from '@components/Tags'
import { colors, getColorByName, siteConfig } from '@config'
import type { Tag, Tags } from '@types'

interface Props {
  tags: Tags
  activeTag?: Tag
}

const TagsCloud = ({ tags, activeTag }: Props): JSX.Element => {
  let tagsList = Object.keys(tags).sort((a, b) => {
    return tags[b] - tags[a]
  })

  if (activeTag) {
    tagsList = tagsList.filter(tag => tag !== activeTag)
  }

  return (
    <div>
      {activeTag ? (
        <LinkTag tag={activeTag} className="md:mb-3 md:text-lg" />
      ) : null}
      {tagsList.map(tag => (
        <LinkTag
          key={tag}
          tag={`${tag} ${tags[tag]}`}
          href={`/tag/${tag}`}
          color={activeTag ? colors.gray : getColorByName(tag)}
          className="md:mb-3 md:text-lg"
        />
      ))}
      {activeTag ? (
        <LinkTag
          tag="All"
          href="/tags"
          color={siteConfig.themeColor}
          className="md:mb-3 md:text-lg"
        />
      ) : null}
    </div>
  )
}

export default TagsCloud
