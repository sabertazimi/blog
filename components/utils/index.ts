import classNames from 'classnames';

export { classNames };
