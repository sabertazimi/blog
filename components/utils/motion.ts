export { AnimatePresence, motion } from 'framer-motion';
export type { MotionProps } from 'framer-motion';
