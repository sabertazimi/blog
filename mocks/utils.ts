let navigatorClipboardMocked = false
const writeTextMock = jest.fn()

function mockNavigatorClipboard() {
  if (!navigatorClipboardMocked) {
    Object.defineProperty(globalThis.navigator, 'clipboard', {
      value: {
        writeText: writeTextMock,
      },
    })

    navigatorClipboardMocked = true
  }

  return { writeTextMock }
}

export { mockNavigatorClipboard }
