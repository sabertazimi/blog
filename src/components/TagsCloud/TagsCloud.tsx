import { Colors, getColorByName } from '@config';
import type { TagType, TagsType } from '@types';
import { Tag } from 'antd';
import { Link } from 'gatsby';
import React from 'react';

interface Props {
  tags: TagsType;
  activeTag: TagType;
}

const TagsCloud = ({ tags, activeTag }: Props): JSX.Element => {
  let tagsList = Object.keys(tags).sort((a, b) => {
    return tags[b] - tags[a];
  });

  if (activeTag) {
    tagsList = tagsList.filter(tag => tag !== activeTag);
  }

  return (
    <div>
      {activeTag ? (
        <Tag className="mb-3" color={getColorByName(activeTag)}>
          <Link className="text-lg font-extrabold" to={`/tags/${activeTag}`}>
            {activeTag} &nbsp; {tags[activeTag]}
          </Link>
        </Tag>
      ) : null}
      {tagsList.map(tag => {
        return (
          <Tag
            key={tag}
            className="mb-3"
            color={activeTag ? Colors.gray : getColorByName(tag)}
          >
            <Link className="text-lg font-extrabold" to={`/tags/${tag}`}>
              {tag} &nbsp; {tags[tag]}
            </Link>
          </Tag>
        );
      })}
      {activeTag ? (
        <Tag className="mb-3 bg-primary border-primary">
          <Link className="text-lg font-extrabold text-light" to="/tags">
            All
          </Link>
        </Tag>
      ) : null}
    </div>
  );
};

export default TagsCloud;
