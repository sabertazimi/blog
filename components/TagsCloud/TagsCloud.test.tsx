import { axe } from 'jest-axe'
import TagsCloud from './TagsCloud'
import { render } from '@/utils'

describe('TagsCloud', () => {
  const mockTag = 'JavaScript'
  const mockTags = {
    'Redux': 1,
    'React': 5,
    'JavaScript': 10,
    'Frontend Development': 20,
    'Web Development': 30,
  }

  it('should render correctly with active tag (snapshot)', () => {
    const { container } = render(
      <TagsCloud activeTag={mockTag} tags={mockTags} />,
    )

    expect(container).toMatchSnapshot()
  })

  it('should render correctly without active tag (snapshot)', () => {
    const { container } = render(<TagsCloud activeTag="" tags={mockTags} />)

    expect(container).toMatchSnapshot()
  })

  it('should render accessibility guidelines (AXE)', async () => {
    const { container } = render(
      <TagsCloud activeTag={mockTag} tags={mockTags} />,
    )

    const a11y = await axe(container)

    expect(a11y).toHaveNoViolations()
  })
})
