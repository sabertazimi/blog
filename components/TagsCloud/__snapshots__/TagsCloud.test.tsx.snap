// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tagsCloud should render correctly with active tag (snapshot) 1`] = `
<div>
  <div>
    <span
      class="ant-tag ant-tag-has-color link dark:hover:shadow-primary md:mb-3 md:text-lg"
      style="background-color: rgb(245, 159, 0);"
    >
      <a
        href="/tag/JavaScript"
      >
        JavaScript
      </a>
    </span>
    <span
      class="ant-tag ant-tag-has-color link dark:hover:shadow-primary md:mb-3 md:text-lg"
      style="background-color: rgb(118, 118, 118);"
    >
      <a
        href="/tag/Web Development"
      >
        Web Development 30
      </a>
    </span>
    <span
      class="ant-tag ant-tag-has-color link dark:hover:shadow-primary md:mb-3 md:text-lg"
      style="background-color: rgb(118, 118, 118);"
    >
      <a
        href="/tag/Frontend Development"
      >
        Frontend Development 20
      </a>
    </span>
    <span
      class="ant-tag ant-tag-has-color link dark:hover:shadow-primary md:mb-3 md:text-lg"
      style="background-color: rgb(118, 118, 118);"
    >
      <a
        href="/tag/React"
      >
        React 5
      </a>
    </span>
    <span
      class="ant-tag ant-tag-has-color link dark:hover:shadow-primary md:mb-3 md:text-lg"
      style="background-color: rgb(118, 118, 118);"
    >
      <a
        href="/tag/Redux"
      >
        Redux 1
      </a>
    </span>
    <span
      class="ant-tag ant-tag-has-color link dark:hover:shadow-primary md:mb-3 md:text-lg"
      style="background-color: rgb(22, 119, 255);"
    >
      <a
        href="/tags"
      >
        All
      </a>
    </span>
  </div>
</div>
`;

exports[`tagsCloud should render correctly without active tag (snapshot) 1`] = `
<div>
  <div>
    <span
      class="ant-tag ant-tag-has-color link dark:hover:shadow-primary md:mb-3 md:text-lg"
      style="background-color: rgb(112, 72, 232);"
    >
      <a
        href="/tag/Web Development"
      >
        Web Development 30
      </a>
    </span>
    <span
      class="ant-tag ant-tag-has-color link dark:hover:shadow-primary md:mb-3 md:text-lg"
      style="background-color: rgb(55, 178, 77);"
    >
      <a
        href="/tag/Frontend Development"
      >
        Frontend Development 20
      </a>
    </span>
    <span
      class="ant-tag ant-tag-has-color link dark:hover:shadow-primary md:mb-3 md:text-lg"
      style="background-color: rgb(245, 159, 0);"
    >
      <a
        href="/tag/JavaScript"
      >
        JavaScript 10
      </a>
    </span>
    <span
      class="ant-tag ant-tag-has-color link dark:hover:shadow-primary md:mb-3 md:text-lg"
      style="background-color: rgb(22, 119, 255);"
    >
      <a
        href="/tag/React"
      >
        React 5
      </a>
    </span>
    <span
      class="ant-tag ant-tag-has-color link dark:hover:shadow-primary md:mb-3 md:text-lg"
      style="background-color: rgb(22, 119, 255);"
    >
      <a
        href="/tag/Redux"
      >
        Redux 1
      </a>
    </span>
  </div>
</div>
`;
